server/lib/scheduler/Scheduler.ts

```diff
@@ -0,0 +1,110 @@
+import { filter } from "asyncro";
+
+class Scheduler {
+  private fetchAllWorkspaces;
+  private rootParentCache;
+  private schedule;
+
+  public constructor({ fetchAllWorkspaces, rootParentCache, schedule }) {
+    this.fetchAllWorkspaces = fetchAllWorkspaces;
+    this.rootParentCache = rootParentCache;
+    this.schedule = schedule;
+  }
+
+  public async getIdOfCurrentWorkspace(userId) {
+    const assignment = this.schedule.getMostRecentAssignmentForUser(userId);
+    return assignment.getWorkspaceId();
+  }
+
+  public async findNextWorkspace(userId) {
+    // clear cache so we don't use old eligibility info
+    this.rootParentCache.clearRootParentCache();
+    const idsOfWorkspacesThatCouldBeNext = await this.getIdsOfWorkspacesThatCouldBeNext(userId);
+    const assignedWorkspaceId = this.pickWorkspaceIdAtRandom(idsOfWorkspacesThatCouldBeNext);
+    await this.schedule.assignWorkspaceToUser(userId, assignedWorkspaceId);
+  }
+
+  private async getIdsOfWorkspacesThatCouldBeNext(userId) {
+    const allWorkspaces = await this.fetchAllWorkspaces();
```

Comment:
Fetching all workspaces and then filtering seems quite inefficient. Any reason not to first find the tree that the user worked on least recently, then only request the workspaces in that tree (and, once time budgets are in, only the ones with remaining budget)?

I guess you'd need to switch to the tree worked on second-least-recently if there are no eligible workspaces in the least-recent one, but you have that issue with the current implementation as well.

Reply:
Addressed here: https://github.com/oughtinc/mosaic/pull/196/commits/a8f7044f0a2107ea1db785a72e503a07b3e690c2.  Instead of asking for `fetchAllWorkspaces`, Scheduler now asks for `fetchAllRootWorkspaces`, and `fetchAllWorkspacesInTree`, and avoids performing a query wrt any workspaces not in the least recently worked on trees.